onnxruntime/contrib_ops/cuda/bert/cutlass_fmha/memory_efficient_attention.cu

```diff
@@ -0,0 +1,152 @@
+    p.head_dim = params.qk_head_size;
+    p.head_dim_value = params.v_head_size;
+
+    // When params.cu_seqlens_q is provided, num_queries is max_seq_q and num_keys will be set inside the kernel
+    p.num_queries = params.sequence_length;
+    p.num_keys = params.kv_sequence_length;
+
+    p.causal = params.causal;
+
+    // Input format is BxSxNxH, output is BxSxNxH
+    p.q_strideH = params.qk_head_size;
+    p.k_strideH = params.qk_head_size;
+    p.v_strideH = params.v_head_size;
+    p.o_strideH = params.v_head_size;
+
+    p.q_strideM = params.num_heads * params.qk_head_size;
+    p.k_strideM = params.num_heads * params.qk_head_size;
+    p.v_strideM = params.num_heads * params.v_head_size;
+
+    p.q_strideB = static_cast<int64_t>(p.q_strideM) * params.sequence_length;
+    p.k_strideB = static_cast<int64_t>(p.k_strideM) * params.kv_sequence_length;
+    p.v_strideB = static_cast<int64_t>(p.v_strideM) * params.kv_sequence_length;
+    p.o_strideB = static_cast<int64_t>(params.num_heads) * params.v_head_size * params.sequence_length;
+
+    p.causal = params.causal;
+  }
+
+  constexpr auto kernel_fn = attention_kernel_batched_impl<Attention>;
+  int smem_bytes = sizeof(typename Attention::SharedStorage);
+  if (smem_bytes > 0xc000) {
+    ORT_ENFORCE(params.sm >= 70, "This kernel requires too much shared memory on this machine!");
+    static bool once = [&]() {
+      cudaFuncSetAttribute(kernel_fn, cudaFuncAttributeMaxDynamicSharedMemorySize, smem_bytes);
+      return true;
+    }();
+  }
+
+  ORT_ENFORCE(Attention::check_supported(p));
+  kernel_fn<<<p.getBlocksGrid(), p.getThreadsGrid(), smem_bytes, params.stream>>>(p);
+}
+
+template <typename T, typename ArchTag, int queries_per_block, int keys_per_block, bool single_value_iteration>
+void DispatchIsAligned(const MemoryEfficientAttentionParams& params) {
+  using AlignedAK = AttentionKernel<T, ArchTag, true, queries_per_block, keys_per_block, single_value_iteration>;
+
+  // Run a more efficient kernel with `isAligned=True` when memory is correctly aligned.
+  bool is_aligned = params.qk_head_size % AlignedAK::kAlignmentQ == 0 &&
+                    params.qk_head_size % AlignedAK::kAlignmentK == 0 &&
+                    params.v_head_size % AlignedAK::kAlignmentV == 0;
+
+  DISPATCH_BOOL(is_aligned, kIsAligned, ([&]() {
+                  LaunchCutlassFmha<T, ArchTag, kIsAligned, queries_per_block, keys_per_block, single_value_iteration>(params);
+                }));
+}
+
+template <typename T, typename ArchTag>
+void DispatchBlockSize(const MemoryEfficientAttentionParams& params) {
+  if (params.v_head_size <= 64) {
+    DispatchIsAligned<T, ArchTag, 64, 64, true>(params);
+  } else if (params.v_head_size <= 128) {
+    DispatchIsAligned<T, ArchTag, 32, 128, true>(params);
+  } else {
+    DispatchIsAligned<T, ArchTag, 32, 128, false>(params);
+  }
+}
+
+template <typename T>
+void DispatchArchTag(const MemoryEfficientAttentionParams& params) {
```

Comment:
nit: inline

Reply:
Let compile decide it. The DispatchBlockSize might be expanded to be a lot of code. inline might not help perf.